common/src/main/scala/com/gu/recipeasy/models/RecipeReadiness.scala

```diff
@@ -11,17 +19,142 @@ object RecipeReadiness {
     recipes.foreach(recipe => db.setOriginalRecipeStatus(recipe.id, Ready))
     recipes.size
   }
+
+  // -------------------------------------------------
+  // Stage 2
+
   def selectNigelSlaterRecipes(db: DB): List[Recipe] = {
-    db.getOriginalRecipes().filter(_.credit.exists(_.contains("Nigel Slater")))
+    db.getOriginalRecipes().filter(_.status != Impossible).filter(_.credit.exists(_.contains("Nigel Slater")))
   }
   def migrateNigelSlaterRecipes(db: DB): Int = {
     val recipes: List[Recipe] = selectNigelSlaterRecipes(db).filter(_.status != Impossible)
     recipes.foreach(recipe => db.setOriginalRecipeStatus(recipe.id, Impossible))
     recipes.size
   }
+
+  // -------------------------------------------------
+  // Stage 3 Support Functions
+
+  def recipeHasACuratedRecordWithNonEmptyIngredientsLists(db: DB, recipe: Recipe): Boolean = {
+    val maybeCuratedRecipe: Option[CuratedRecipeDB] = db.getCuratedRecipeByRecipeId(recipe.id)
+    maybeCuratedRecipe.exists(curatedRecipe => curatedRecipe.ingredientsLists.lists.size > 0)
+  }
+
+  private object ShortListItem {
+    private def isShort(length: Int) = length > 0 && length < 120
+    def matches(node: Node): Boolean = node match {
+      case tn: TextNode => isShort(tn.text.trim.size)
+      case elem: Element if (elem.tag.getName == "strong" || elem.tag.getName == "em") => { isShort(elem.text.trim.size) }
+      case other => false
+    }
+  }
+
+  private object ParaWithListOfShortTexts {
+    def matches(el: Element): Boolean = {
+      if (el == null) false
+      else {
+        if (el.tag.getName == "p") {
+          val pairs = el.childNodes.asScala.toList.grouped(2)
+          pairs.forall {
+            case x :: y :: Nil => ShortListItem.matches(x) && y.nodeName == "br"
+            case x :: Nil => ShortListItem.matches(x) // final element
+            case _ => false
+          }
+        } else false
+      }
+    }
+    def unapply(el: Element): Option[Element] = if (matches(el)) Some(el) else None
+  }
+  def buildIngredientList(para: Element): Option[IngredientsList] = {
+    def text(node: Node): String = node match {
+      case tn: TextNode => tn.text.trim
+      case elem: Element => elem.text.trim
+      case other => ""
+    }
+    val listItems: Seq[Node] = para.childNodes.asScala.filterNot(_.nodeName == "br")
+    val withoutServingCount = listItems.filterNot(n => text(n).toLowerCase.dropWhile(_ == '(').startsWith("serves"))
+    if (withoutServingCount.size < 2) None
+    else {
+      if (withoutServingCount(0).nodeName != withoutServingCount(1).nodeName) {
```

Comment:
The only thing protecting you from an `indexOutOfBoundsException` here if the conditional statement which may in time change, so it's best to access the array elements in a way that gives you back an Option.